Library/Line.cpp

```diff
@@ -0,0 +1,227 @@
+﻿#include "Line.h"
+
+#include <fstream>
+
+using namespace Library;
+using namespace std;
+
+Line::Line(string name)
+{
+	m_name = name;
+}
+
+Line::Line(const Line& L)
+{
+	m_name = L.m_name;
+	try
+	{
+		for (auto s : L.m_stations) //клонируем каждый элемент из контейнера другой линии
+			m_stations.push_back(s->clone());
+	}
+	catch (const exception& e) //если при копировании возникло исключение
+	{
+		for (auto s : m_stations) //s имеет тип Station*
+			delete s; //очищаем что накопировали
+		throw e; //и выкидываем исключение дальше
+	}
+}
+
+Line::Line(Line&& L)
+{
+	m_name = L.m_name;
+	m_stations = L.m_stations;
+	L.m_stations.clear();
+}
+
+Line::~Line()
+{
+	for(auto s : m_stations)
+		delete s; //удаляем все станции в контейнере
+}
+
+Line& Line::operator= (const Line& L)
+{
+	if (this != &L)
+	{
+		m_name = L.m_name;
+		deque<Station*> stations;
+		try
+		{
+			for (auto s : L.m_stations) //клонируем каждый элемент из контейнера другой линии
+				stations.push_back(s->clone());
+			for (auto s : m_stations)
+				delete s; //удаляем все станции, которые хранились
+			m_stations.clear();
+			m_stations = stations;
+		}
+		catch (const exception& e) //если при копировании возникло исключение
+		{
+			for (auto s : stations) //s - имеет тип Station*
+				delete s; //очищаем что накопировали
+			throw e; //и выкидываем исключение дальше
+		}
+	}
+	return *this;
+}
+
+Line& Line::operator= (Line&& L)
+{
+	if (this != &L)
+	{
+		m_name = L.m_name;
+		m_stations = L.m_stations;
+		L.m_stations.clear();
+	}
+	return *this;
+}
+
+
+std::ostream& Library::operator<< (std::ostream& o, const Line& L)
+{
+	o << "Line: " << L.m_name << endl << endl;
```

Comment:
endl имеет больший функционал чем просто перенос строки. Пока не придет endl, не будет вызван метод flush, или не будет записано определенное количество элементов, поток будет накапливать в буфере все символы не печатая их. Когда приходит endl поток сразу же напечатает все что успел накопить. Это сделано так потому что так гораздо быстрее. Когда вы два раза подряд передаете endl вы по факту заставляете его распечатать один (или 2) символ переноса строки без буферизации, что потратит больше времени чем должно. Лучше либо так:
```suggestion
	o << "Line: " << L.m_name << '\n' << endl;
```
Либо вообще все endl заменить на \`\n\` и в конце метода вызвать o.flush()